integration-cli/docker_cli_build_test.go

```diff
@@ -762,41 +762,188 @@ RUN [ $(ls -l /exists/exists_file | awk '{print $3":"$4}') = 'dockerio:dockerio'
+	defer deleteImages(name)
+
+	testAddFn := func(t *testing.T, instruction string) {
+		ctx, err := fakeContext(
+			fmt.Sprintf("FROM scratch\n%s\n", instruction),
+			map[string]string{
+				"file1.txt": "test1",
+				"file2.txt": "test1",
+			})
+		defer ctx.Close()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		expected := "When using ADD with more than one source file, the destination must be a directory and end with a /"
+		if _, err := buildImageFromContext(name, ctx, true); err == nil || !strings.Contains(err.Error(), expected) {
+			t.Fatalf("Wrong error: (should contain %q) got:\n%v", expected, err)
+		}
+	}
+
+	for _, addInstruction := range []string{
+		`ADD file*.txt test`,
+		`ADD ["file*.txt", "test"]`,
+	} {
+		testAddFn(t, addInstruction)
+	}
+
+	logDone("build - multiple add files to file wild")
+}
+
+func TestBuildCopyMultipleFilesToFile(t *testing.T) {
+	name := "testcopymultiplefilestofile"
+	defer deleteImages(name)
+
+	testCopyFn := func(t *testing.T, instruction string) {
+		ctx, err := fakeContext(
+			fmt.Sprintf("FROM scratch\n%s\n", instruction),
+			map[string]string{
+				"file1.txt": "test1",
+				"file2.txt": "test1",
+			})
+		defer ctx.Close()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		expected := "When using COPY with more than one source file, the destination must be a directory and end with a /"
+		if _, err := buildImageFromContext(name, ctx, true); err == nil || !strings.Contains(err.Error(), expected) {
+			t.Fatalf("Wrong error: (should contain %q) got:\n%v", expected, err)
+		}
+	}
+
+	for _, copyInstruction := range []string{
+		`COPY file1.txt file2.txt test`,
+		`COPY ["file1.txt", "file2.txt", "test"]`,
+	} {
+		testCopyFn(t, copyInstruction)
+	}
+
+	logDone("build - multiple copy files to file")
+}
+
+func TestBuildAddFileWithWhitespace(t *testing.T) {
+	name := "testaddfilewithwhitespace"
+	defer deleteImages(name)
+	ctx, err := fakeContext(`FROM busybox
+RUN mkdir "/test dir"
+RUN mkdir "/test_dir"
+ADD [ "test file1", "/test_file1" ]
+ADD [ "test_file2", "/test file2" ]
+ADD [ "test file3", "/test file3" ]
+ADD [ "test dir/test_file4", "/test_dir/test_file4" ]
+ADD [ "test_dir/test_file5", "/test dir/test_file5" ]
+ADD [ "test dir/test_file6", "/test dir/test_file6" ]
+RUN [ $(cat "/test_file1") = 'test1' ]
+RUN [ $(cat "/test file2") = 'test2' ]
+RUN [ $(cat "/test file3") = 'test3' ]
+RUN [ $(cat "/test_dir/test_file4") = 'test4' ]
+RUN [ $(cat "/test dir/test_file5") = 'test5' ]
+RUN [ $(cat "/test dir/test_file6") = 'test6' ]`,
 		map[string]string{
-			"file1.txt": "test1",
-			"file2.txt": "test1",
+			"test file1":          "test1",
+			"test_file2":          "test2",
+			"test file3":          "test3",
+			"test dir/test_file4": "test4",
+			"test_dir/test_file5": "test5",
+			"test dir/test_file6": "test6",
 		})
 	defer ctx.Close()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	expected := "When using ADD with more than one source file, the destination must be a directory and end with a /"
```

Comment:
I think the diff view is misleading: you can find it on line 784 of the patched file.